client/driver/executor/executor_libcontainer.go

```diff
@@ -0,0 +1,512 @@
+	}
+}
+
+// LibcontainerExecutor implements an Executor with the runc/libcontainer api
+type LibcontainerExecutor struct {
+	id      string
+	command *ExecCommand
+
+	logger hclog.Logger
+
+	totalCpuStats  *stats.CpuStats
+	userCpuStats   *stats.CpuStats
+	systemCpuStats *stats.CpuStats
+	pidCollector   *pidCollector
+
+	container      libcontainer.Container
+	userProc       *libcontainer.Process
+	userProcExited chan interface{}
+	exitState      *ProcessState
+}
+
+func NewExecutorWithIsolation(logger hclog.Logger) Executor {
+	logger = logger.Named("executor")
+	if err := shelpers.Init(); err != nil {
+		logger.Error("unable to initialize stats", "error", err)
+	}
+	return &LibcontainerExecutor{
+		id:             strings.Replace(uuid.Generate(), "-", "_", 0),
+		logger:         logger,
+		totalCpuStats:  stats.NewCpuStats(),
+		userCpuStats:   stats.NewCpuStats(),
+		systemCpuStats: stats.NewCpuStats(),
+		pidCollector:   newPidCollector(logger),
+	}
+}
+
+// Launch creates a new container in libcontainer and starts a new process with it
+func (l *LibcontainerExecutor) Launch(command *ExecCommand) (*ProcessState, error) {
+	// Find the nomad executable to launch the executor process with
+	bin, err := discover.NomadExecutable()
+	if err != nil {
+		return nil, fmt.Errorf("unable to find the nomad binary: %v", err)
+	}
+
+	l.command = command
+
+	// create a new factory which will store the container state in the allocDir
+	factory, err := libcontainer.New(
+		path.Join(command.TaskDir, "../alloc/container"),
+		libcontainer.Cgroupfs,
+		libcontainer.InitArgs(bin, "libcontainer-shim"),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create factory: %v", err)
+	}
+
+	// A container groups processes under the same isolation enforcement
+	container, err := factory.Create(l.id, newLibcontainerConfig(command))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create container(%s): %v", l.id, err)
+	}
+	l.container = container
+
+	combined := append([]string{command.Cmd}, command.Args...)
+	stdout, err := command.Stdout()
+	if err != nil {
+		return nil, err
+	}
+	stderr, err := command.Stderr()
+	if err != nil {
+		return nil, err
+	}
+
+	// the task process will be started by the container
+	process := &libcontainer.Process{
+		Args:   combined,
+		Env:    command.Env,
+		Stdout: stdout,
+		Stderr: stderr,
+		Init:   true,
+	}
+	l.userProc = process
+
+	l.totalCpuStats = stats.NewCpuStats()
+	l.userCpuStats = stats.NewCpuStats()
+	l.systemCpuStats = stats.NewCpuStats()
+
+	// Starts the task
+	if err := container.Run(process); err != nil {
+		container.Destroy()
+		return nil, err
+	}
+
+	pid, err := process.Pid()
+	if err != nil {
+		container.Destroy()
+		return nil, err
+	}
+
+	// start a goroutine to wait on the process to complete, so Wait calls can
+	// be multiplexed
+	l.userProcExited = make(chan interface{})
+	go l.pidCollector.collectPids(l.userProcExited)
```

Comment:
As mentioned we can probably collect these pids more efficiently using the cgroup (libcontainer interface)